### PR TITLE
Feature/abuse protection config

### LINKED ISSUE #120: Manage abuse protection
Currently all abuse protection mechanisms are always enabled.

It should be possible to disable and enable every single mechanism via an own command.

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
Allow to manage the abuse protection settings.

- [x] Rework guild settings
- [x] Cleanup and reogranize database tables
- [x] Split repsettings into repsettings and abuse protection
- [x] Add embed toggle to rep settings to disable the confirmation dialog
- [x] Add donor and receiver context to abuse protection settings

Sorry for the size. This should have been 3 PRs

Close #120

### CODE REVIEW

File: src/main/java/de/chojo/repbot/commands/AbuseProtection.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+            return minMessages(eventWrapper, context.subContext(subcmd), guildSettings);
+        }
+
+        if ("cooldown".equalsIgnoreCase(subcmd)) {
+            return cooldown(eventWrapper, context.subContext(subcmd), guildSettings);
+        }
+        if ("donorContext".equalsIgnoreCase(subcmd)) {
+            return donorContext(eventWrapper, context.subContext(subcmd), guildSettings);
+        }
+        if ("receiverContext".equalsIgnoreCase(subcmd)) {
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/commands/AbuseProtection.java:87]:
```suggestion
        if ("cooldown".equalsIgnoreCase(subcmd)) {
            return cooldown(eventWrapper, context.subContext(subcmd), guildSettings);
        }
        
        if ("donorContext".equalsIgnoreCase(subcmd)) {
            return donorContext(eventWrapper, context.subContext(subcmd), guildSettings);
        }
        
        if ("receiverContext".equalsIgnoreCase(subcmd)) {
```

File: src/main/java/de/chojo/repbot/commands/AbuseProtection.java [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+            minMessages(event, guildSettings);
+        }
+
+        if ("cooldown".equalsIgnoreCase(subcmd)) {
+            cooldown(event, guildSettings);
+        }
+        if ("donorContext".equalsIgnoreCase(subcmd)) {
+            donorContext(event, guildSettings);
+        }
+        if ("receiverContext".equalsIgnoreCase(subcmd)) {
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/commands/AbuseProtection.java:116]:
```suggestion
        }
        
        if ("donorContext".equalsIgnoreCase(subcmd)) {
            donorContext(event, guildSettings);
        }
        
        if ("receiverContext".equalsIgnoreCase(subcmd)) {
```

File: src/main/java/de/chojo/repbot/data/wrapper/AbuseSettings.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        this.cooldown = cooldown;
+        this.maxMessageAge = maxMessageAge;
+        this.minMessages = minMessages;
+        this.donorContext = donorContext;
+        this.receiverContext = receiverContext;
+    }
+
+    public AbuseSettings() {
+    }
+
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/data/wrapper/AbuseSettings.java:25]:
```suggestion
```

File: src/main/java/de/chojo/repbot/data/wrapper/GeneralSettings.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+public class GeneralSettings {
+    private final String prefix;
+    private final boolean emojiDebug;
+    private final long managerRole;
+
+    public GeneralSettings() {
+        prefix = null;
+        emojiDebug = true;
+        managerRole =0;
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/data/wrapper/GeneralSettings.java:17]:
```suggestion
        managerRole = 0;
```

File: src/main/java/de/chojo/repbot/data/wrapper/ThankSettings.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    private final boolean channelWhitelist;
+    private final Set<Long> donorRoles;
+    private final Set<Long> receiverRoles;
+
+    public ThankSettings() {
+        reaction = null;
+        reactions = Collections.emptySet();
+        thankwords = new String[0];
+        activeChannel = Collections.emptySet();
+        channelWhitelist =true;
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/data/wrapper/ThankSettings.java:36]:
```suggestion
        channelWhitelist = true;
```

File: src/main/resources/locale_de_DE.properties [hunk 6]
```diff
@@ -1,3 +1,24 @@
...
+command.abuseProtection.sub.donorContext.false=Der Spender braucht keine aktuellen Nachrichten im Kanal.
+command.abuseProtection.sub.donorContext.true=Der Spender muss aktuelle Nachrichten im Kanal haben.
+command.abuseProtection.sub.info=Abrufen der aktuellen Einstellungen.
+command.abuseProtection.sub.maxMessageAge=Zeigt oder ändert das maximale Nachrichtenalter für Reputation.
+command.abuseProtection.sub.maxMessageAge.get=Nachrichten, die älter als %MINUTES% Minuten sind, können keine Reputation erhalten.
+command.abuseProtection.sub.minMessages=Der Nutzer kann den Verfassern dieser Nachrichten immer danken.
+command.abuseProtection.sub.minMessages.get=Der Nutzer kann den Verfassern der letzten %AMOUNT% Nachrichten immer danken.
+command.abuseProtection.sub.receiverContext=Lege fest ob der Empfänger aktuelle Nachrichten benötigt oder nicht.
+command.abuseProtection.sub.receiverContext.false=Der Empfänger benötigt keine aktuellen Nachrichten im Kanal.
+command.abuseProtection.sub.receiverContext.true=Der Empfänger muss aktuelle Nachrichten im Kanal haben.
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:21]:
```suggestion
command.abuseProtection.description=Zeigt die aktuellen Missbrauchsschutzeinstellungen an.
command.abuseProtection.embed.descr.cooldown=Cooldown
command.abuseProtection.embed.descr.donorContext=Geberkontext
command.abuseProtection.embed.descr.maxMessageAge=Max. Nachrichtenalter
command.abuseProtection.embed.descr.minMessages=Min. Nachrichten
command.abuseProtection.embed.descr.receiverContext=Spenderkontext
command.abuseProtection.embed.title=Missbrauchsschutz
command.abuseProtection.sub.cooldown=Zeigt oder ändert die Abklingzeit für die Reputation.
command.abuseProtection.sub.cooldown.get=Man kann nun aller %MINUTES% Minuten einem Nutzer Reputation geben.
command.abuseProtection.sub.cooldown.set=Abklingzeit auf %MINUTES% Minuten gesetzt.
command.abuseProtection.sub.donorContext=Legt fest, ob der Spender aktuelle Nachrichten in dem Channel benötigt.
command.abuseProtection.sub.donorContext.false=Der Spender braucht keine aktuellen Nachrichten im Kanal.
command.abuseProtection.sub.donorContext.true=Der Spender muss aktuelle Nachrichten im Kanal haben.
command.abuseProtection.sub.info=Zeigt die aktuellen Einstellungen an.
command.abuseProtection.sub.maxMessageAge=Zeigt oder ändert das maximale Nachrichtenalter für Reputation.
command.abuseProtection.sub.maxMessageAge.get=Nachrichten, die älter als %MINUTES% Minuten sind, können keine Reputation erhalten.
command.abuseProtection.sub.minMessages=Der Nutzer kann den Verfassern dieser Nachrichten immer danken.
command.abuseProtection.sub.minMessages.get=Der Nutzer kann den Verfassern der letzten %AMOUNT% Nachrichten immer danken.
command.abuseProtection.sub.receiverContext=Legt fest, ob der Empfänger aktuelle Nachrichten benötigt oder nicht.
command.abuseProtection.sub.receiverContext.false=Der Empfänger benötigt keine aktuellen Nachrichten im Kanal.
command.abuseProtection.sub.receiverContext.true=Der Empfänger muss aktuelle Nachrichten im Kanal haben.
```

**Review by dlsf — CHANGES_REQUESTED** [review]

File: src/main/java/de/chojo/repbot/data/wrapper/AbuseSettings.java (hunk 3 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/data/wrapper/AbuseSettings.java:25]: this constructor should contain reasonable default values. Gonna change this.

File: src/main/java/de/chojo/repbot/commands/RepSettings.java [hunk 7]
```diff
@@ -331,129 +267,96 @@ private boolean answer(MessageEventWrapper eventWrapper, CommandContext context,
...
         if (event.getOptions().isEmpty()) {
-            event.reply(getBooleanMessage(event.getGuild(), guildSettings.isAnswerActive(),
+            event.reply(getBooleanMessage(event.getGuild(), guildSettings.messageSettings().isAnswerActive(),
                     "command.repSettings.sub.answer.true", "command.repSettings.sub.answer.false")).queue();
             return;
         }
         var answer = event.getOption("answer").getAsBoolean();
 
-        if (data.updateMessageSettings(GuildSettingUpdate.builder(event.getGuild()).answerActive(answer).build())) {
+        guildSettings.messageSettings();
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/RepSettings.java:276]: Result of method call is ignored and this method does not have side effects

File: src/main/java/de/chojo/repbot/commands/RepSettings.java [hunk 8]
```diff
@@ -331,129 +267,96 @@ private boolean answer(MessageEventWrapper eventWrapper, CommandContext context,
...
+        if (optEmbed.isEmpty()) {
             eventWrapper.replyErrorAndDelete(eventWrapper.localize("error.notABoolean",
                     Replacement.create("INPUT", context.argString(0))), 30);
             return false;
         }
 
-        if (data.updateMessageSettings(GuildSettingUpdate.builder(eventWrapper.getGuild()).reactionsActive(optReactions.get()).build())) {
-            eventWrapper.reply(getBooleanMessage(eventWrapper.getGuild(), optReactions.get(),
-                    "command.repSettings.sub.reactions.true", "command.repSettings.sub.reactions.false")).queue();
+        guildSettings.messageSettings().embedActive(optEmbed.get());
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/RepSettings.java:297]: Result of method call is ignored and this method does not have side effects

File: src/main/java/de/chojo/repbot/commands/RepSettings.java [hunk 9]
```diff
@@ -331,129 +267,96 @@ private boolean answer(MessageEventWrapper eventWrapper, CommandContext context,
...
+                    "command.repSettings.sub.embed.true", "command.repSettings.sub.embed.false")).queue();
             return;
         }
-        var reactions = event.getOption("reactions").getAsBoolean();
+        var embed = event.getOption("embed").getAsBoolean();
 
-        if (data.updateMessageSettings(GuildSettingUpdate.builder(event.getGuild()).reactionsActive(reactions).build())) {
-            event.reply(getBooleanMessage(event.getGuild(), reactions,
-                    "command.repSettings.sub.reactions.true", "command.repSettings.sub.reactions.false")).queue();
+        guildSettings.messageSettings().embedActive(embed);
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/RepSettings.java:313]: Result of method call is ignored and this method does not have side effects

File: src/main/java/de/chojo/repbot/commands/RepSettings.java [hunk 10]
```diff
@@ -331,129 +267,96 @@ private boolean answer(MessageEventWrapper eventWrapper, CommandContext context,
...
 
-        if (optAge.isEmpty()) {
-            eventWrapper.replyErrorAndDelete(context.argString(0).get() + " is not a number", 30);
-            return true;
-        }
-        Integer minMessages = Math.max(0, Math.min(optAge.get(), 100));
-        if (data.updateMessageSettings(GuildSettingUpdate.builder(eventWrapper.getGuild()).minMessages(minMessages).build())) {
-            eventWrapper.reply(eventWrapper.localize("command.repSettings.sub.minMessages.get",
-                    Replacement.create("AMOUNT", minMessages))).queue();
+        guildSettings.messageSettings();
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/RepSettings.java:334]: Result of method call is ignored and this method does not have side effects

File: src/main/java/de/chojo/repbot/commands/RepSettings.java [hunk 11]
```diff
@@ -331,129 +267,96 @@ private boolean answer(MessageEventWrapper eventWrapper, CommandContext context,
...
             return;
         }
-        var minMessages = event.getOption("messages").getAsLong();
+        var reactions = event.getOption("reactions").getAsBoolean();
 
-        minMessages = Math.max(0, Math.min(minMessages, 100));
-        if (data.updateMessageSettings(GuildSettingUpdate.builder(event.getGuild()).minMessages((int) minMessages).build())) {
-            event.reply(loc.localize("command.repSettings.sub.minMessages.get",
-                    Replacement.create("AMOUNT", minMessages))).queue();
+        guildSettings.messageSettings();
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/commands/RepSettings.java:350]: Result of method call is ignored and this method does not have side effects

File: src/main/resources/locale_de_DE.properties [hunk 12]
```diff
@@ -223,6 +241,10 @@ listener.reaction.confirmation=%DONOR% hat %RECEIVER% Reputation gegeben.
 listener.reaction.removal=Reputation von %DONOR% wurde zurückgenommen.
 message.welcome=Danke, dass du mich eingeladen hast.\nUm direkt loszulegen, kannst du mit dem /setup Befehl die grundlegenden Eintellungen vornehmen.
 messages.destruction=Diese Nachricht zerstört sich in %MIN% Minute(n)
+selfCleanup.leave=Du hast mich nicht konfiguriert.\nIch bleibe nur in aktiven Gilden, die mindestens einen Rufkanal haben. Wenn du mich wieder benutzen willst, kannst du ihn jederzeit wieder mit diesem [Link](%INVITE%) einladen.
```
Comment (Taucher2003) [inline src/main/resources/locale_de_DE.properties:244]:
```suggestion
selfCleanup.leave=Du hast mich nicht konfiguriert.\nIch bleibe nur in aktiven Gilden, die mindestens einen Reputationskanal haben. Wenn du mich wieder benutzen willst, kannst du ihn jederzeit wieder mit diesem [Link](%INVITE%) einladen.
```

**Review by Taucher2003 — CHANGES_REQUESTED** [review]

File: src/main/java/de/chojo/repbot/commands/RepSettings.java [hunk 13]
```diff
@@ -251,7 +182,8 @@ private boolean fuzzy(MessageEventWrapper eventWrapper, CommandContext context,
             return false;
         }
 
-        if (data.updateMessageSettings(GuildSettingUpdate.builder(eventWrapper.getGuild()).fuzzyActive(optFuzzy.get()).build())) {
+        guildSettings.messageSettings();
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/commands/RepSettings.java:185]: Should check my inspections... Don't know why this doesn't show up.

**Review by Taucher2003 — APPROVED** [review]
Looks good to me

**Review by dlsf — APPROVED** [review]

**Pull request merged**